src/server/frontend_wayland/virtual_keyboard_v1.cpp

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright © 2018-2019 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: William Wold <william.wold@canonical.com>
+ */
+
+#include "virtual_keyboard_v1.h"
+#include "virtual-keyboard-unstable-v1_wrapper.h"
+#include "wayland_wrapper.h"
+#include "wayland_wrapper.h"
+
+#include "mir/input/event_builder.h"
+#include "mir/input/input_sink.h"
+#include "mir/input/input_device_info.h"
+#include "mir/input/pointer_settings.h"
+#include "mir/input/touchpad_settings.h"
+#include "mir/input/touchscreen_settings.h"
+#include "mir/input/input_device_registry.h"
+#include "mir/input/input_device.h"
+#include "mir/input/device.h"
+#include "mir/input/mir_keyboard_config.h"
+#include "mir/input/buffer_keymap.h"
+#include "mir/fatal.h"
+#include "mir/log.h"
+
+#include <mutex>
+#include <cstring>
+
+namespace mf = mir::frontend;
+namespace mw = mir::wayland;
+namespace mi = mir::input;
+
+namespace
+{
+auto unique_keyboard_name() -> std::string
+{
+    static std::mutex mutex;
+    static int id = 0;
+    std::lock_guard<std::mutex> lock{mutex};
+    auto const result = "virt-key-" + std::to_string(id);
+    id++;
```

Comment:
This could simply be an atomic `next_id` with `fetch_add(1)` with no need for a mutex.